asllib/aslspec/spec.ml

```diff
@@ -448,16 +448,225 @@ module Check = struct
       List.iter check_prose_template_for_definition_node defining_nodes
   end
 
-  (** TODO:
-      - Check that instantiated type terms for labelled type terms and labelled
-        records match their definition (see [tests/type_instance.bad]). *)
+  (** * A module for conservatively checking that all type terms are
+      well-formed. That is each term that instantiates another defined type term
+      is subsumed by it. *)
+  module CheckTypeInstantiations : sig
+    val check : definition_node StringMap.t -> elem list -> unit
+    (** [check id_to_defining_node elems] conservatively checks that all type
+        terms referenced in [elems] that instantiate type terms in the range of
+        [id_to_defining_node] are also subsumed by them. The check assumes that
+        [check_no_undefined_ids] has already been run. *)
+  end = struct
+    (** [subsumed sub_op super_op] is true if all values in the domain of
+        [Operator { op = sub_op; term }] are also in the domain of
+        [Operator { op = super_op; term }]. *)
+    let operator_subsumed sub_op super_op =
+      match (sub_op, super_op) with
+      | Powerset, Powerset
+      | Powerset_Finite, Powerset_Finite
+      | Powerset_Finite, Powerset
+      | List1, List1
+      | List0, List0
+      | List1, List0
+      | Option, Option
+      | Option, Powerset_Finite
+      | Option, Powerset
```

Comment:
nit: purely stylistic, but I find it more readable/intuitive to write as follows
```suggestion
      | Powerset, Powerset
      | Powerset_Finite, (Powerset_Finite | Powerset)
      | List1, (List1 | List0)
      | List0, List0
      | Option, (Option | Powerset_Finite | Powerset)
```